File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/core/counter/CounterRepositoryTest.java
package coolpharaoh.tee.speicher.tea.timer.core.counter;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;

import coolpharaoh.tee.speicher.tea.timer.database.TeaMemoryDatabase;
import coolpharaoh.tee.speicher.tea.timer.views.export_import.data_transform.pojo.StatisticsPOJO;


@ExtendWith(MockitoExtension.class)
class CounterRepositoryTest {
    @Mock
    TeaMemoryDatabase teaMemoryDatabase;
    @Mock
    CounterDao counterDao;

    private CounterRepository counterRepository;

    @BeforeEach
    void setUp() {
        TeaMemoryDatabase.setMockedDatabase(teaMemoryDatabase);
        when(teaMemoryDatabase.getCounterDao()).thenReturn(counterDao);

        counterRepository = new CounterRepository(null);
    }

    @Test
    void insertCounter() {
        final Counter counter = new Counter();

        counterRepository.insertCounter(counter);

        verify(counterDao).insert(counter);
    }

    @Test
    void updateCounter() {
        final Counter counter = new Counter();

        counterRepository.updateCounter(counter);

        verify(counterDao).update(counter);
    }

    @Test
    void getCounters() {
        when(counterDao.getCounters()).thenReturn(Arrays.asList(new Counter(), new Counter()));

        final List<Counter> counters = counterRepository.getCounters();

        verify(counterDao).getCounters();
        assertThat(counters).hasSize(2);
    }

    @Test
    void getCounterByTeaId() {
        final int teaId = 2;
        final Counter counter = new Counter();
        when(counterDao.getCounterByTeaId(teaId)).thenReturn(counter);

        final Counter counterByTeaId = counterRepository.getCounterByTeaId(teaId);

        assertThat(counterByTeaId).isEqualTo(counter);
    }

    @Test
    void getTeaCounterOverall() {
        when(counterDao.getTeaCounterOverall()).thenReturn(Arrays.asList(new StatisticsPOJO(), new StatisticsPOJO()));

        final List<StatisticsPOJO> counters = counterRepository.getTeaCounterOverall();

        verify(counterDao).getTeaCounterOverall();
        assertThat(counters).hasSize(2);
    }

    @Test
    void getTeaCounterYear() {
        when(counterDao.getTeaCounterYear()).thenReturn(Arrays.asList(new StatisticsPOJO(), new StatisticsPOJO()));

        final List<StatisticsPOJO> counters = counterRepository.getTeaCounterYear();

        verify(counterDao).getTeaCounterYear();
        assertThat(counters).hasSize(2);
    }

    @Test
    void getTeaCounterMonth() {
        when(counterDao.getTeaCounterMonth()).thenReturn(Arrays.asList(new StatisticsPOJO(), new StatisticsPOJO()));

        final List<StatisticsPOJO> counters = counterRepository.getTeaCounterMonth();

        verify(counterDao).getTeaCounterMonth();
        assertThat(counters).hasSize(2);
    }

    @Test
    void getTeaCounterWeek() {
        when(counterDao.getTeaCounterWeek()).thenReturn(Arrays.asList(new StatisticsPOJO(), new StatisticsPOJO()));

        final List<StatisticsPOJO> counters = counterRepository.getTeaCounterWeek();

        verify(counterDao).getTeaCounterWeek();
        assertThat(counters).hasSize(2);
    }
}

File: app/src/test/java/coolpharaoh/tee/speicher/tea/timer/core/counter/CounterRepositoryTest.kt
package coolpharaoh.tee.speicher.tea.timer.core.counter

import android.app.Application
import coolpharaoh.tee.speicher.tea.timer.database.TeaMemoryDatabase
import coolpharaoh.tee.speicher.tea.timer.database.TeaMemoryDatabase.Companion.setMockedDatabase
import coolpharaoh.tee.speicher.tea.timer.views.export_import.data_transform.pojo.StatisticsPOJO
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
internal class CounterRepositoryTest {
    @MockK
    lateinit var teaMemoryDatabase: TeaMemoryDatabase
    @RelaxedMockK
    lateinit var counterDao: CounterDao

    private var counterRepository: CounterRepository? = null

    @BeforeEach
    fun setUp() {
        setMockedDatabase(teaMemoryDatabase)
        every { teaMemoryDatabase.counterDao } returns counterDao

        counterRepository = CounterRepository(Application())
    }

    @Test
    fun insertCounter() {
        val counter = Counter()

        counterRepository!!.insertCounter(counter)

        verify { counterDao.insert(counter) }
    }

    @Test
    fun updateCounter() {
        val counter = Counter()

        counterRepository!!.updateCounter(counter)

        verify { counterDao.update(counter) }
    }

    @Test
    fun counters() {
        every { counterDao.getCounters() } returns listOf(Counter(), Counter())

        val counters = counterRepository!!.counters

        verify { counterDao.getCounters() }
        assertThat(counters).hasSize(2)
    }

    @Test
    fun counterByTeaId() {
        val teaId = 2
        val counter = Counter()
        every { counterDao.getCounterByTeaId(teaId.toLong()) } returns counter

        val counterByTeaId = counterRepository!!.getCounterByTeaId(teaId.toLong())

        assertThat(counterByTeaId).isEqualTo(counter)
    }

    @Test
    fun teaCounterOverall() {
        every { counterDao.getTeaCounterOverall() } returns listOf(StatisticsPOJO(), StatisticsPOJO())

        val counters = counterRepository!!.teaCounterOverall

        verify { counterDao.getTeaCounterOverall() }
        assertThat(counters).hasSize(2)
    }

    @Test
    fun teaCounterYear() {
        every { counterDao.getTeaCounterYear() } returns listOf(StatisticsPOJO(), StatisticsPOJO())

        val counters = counterRepository!!.teaCounterYear

        verify { counterDao.getTeaCounterYear() }
        assertThat(counters).hasSize(2)
    }

    @Test
    fun teaCounterMonth() {
        every { counterDao.getTeaCounterMonth() } returns listOf(StatisticsPOJO(), StatisticsPOJO())

        val counters = counterRepository!!.teaCounterMonth

        verify { counterDao.getTeaCounterMonth() }
        assertThat(counters).hasSize(2)
    }

    @Test
    fun teaCounterWeek() {
        every { counterDao.getTeaCounterWeek() } returns listOf(StatisticsPOJO(), StatisticsPOJO())

        val counters = counterRepository!!.teaCounterWeek

        verify { counterDao.getTeaCounterWeek() }
        assertThat(counters).hasSize(2)
    }
}